src/main/groovy/org/labkey/gradle/task/TeamCityPropertiesTask.groovy

```diff
@@ -0,0 +1,147 @@
+            tcPropOrDefault(project,
+                    TeamCityPropertiesTask::getLabKeyServerShutdownPort,
+                    "shutdownPort",
+                    "8081"))
+    @Optional @Input
+    final abstract Property<String> keyStore = project.objects.property(String).convention(
+            tcPropOrDefault(project,
+                    TeamCityPropertiesTask::getLabKeyServerKeystore,
+                    "keyStore",
+                    "/opt/teamcity-agent/localhost.keystore"))
+
+    @Optional @Input
+    final abstract Property<String> keyStorePassword = project.objects.property(String).convention(
+            tcPropOrDefault(project,
+                    TeamCityPropertiesTask::getLabKeyServerKeystorePassword,
+                    "keyStorePassword",
+                    "changeit"))
+
+    @Optional @Input
+    final abstract Property<String> labKeyServer = project.objects.property(String).convention(getLabKeyServer(project))
+
+    @Optional @Input
+    final abstract Property<String> tomcatJavaHome = project.objects.property(String).convention(
+            tcPropOrDefault(project,
+                TeamCityPropertiesTask::getTomcatJavaHome,
+                    "tomcatJavaHome",
+                    System.getenv("JAVA_HOME")
+            )
+    )
+
+    protected static String tcPropOrDefault(Project project, Function<Project, String> tcPropertyFunc, String projectPropertyName, String defaultValue)
+    {
+        String value = tcPropertyFunc.apply(project)
+        if (value == null) {
+            if (project.hasProperty(projectPropertyName))
+                value = (String) project.property(projectPropertyName)
+            else
+                value = defaultValue
+        }
+        return value
+    }
+
+    static boolean isOnTeamCity(Project project)
+    {
+        return project.hasProperty('teamcity')
+    }
+
+    static Object getTeamCityProperty(Project project, String name, Object defaultValue)
+    {
+        if (isOnTeamCity(project))
+            return project.teamcity[name] != null ? project.teamcity[name] : defaultValue
+        else if (project.hasProperty(name))
+            return project.property(name)
+        else
+            return defaultValue
+    }
+
+    static Properties getTeamCityProperties(Project project)
+    {
+        if (isOnTeamCity(project))
+        {
+            def tcProps = new Properties()
+            tcProps.putAll(project.teamcity)
+            return tcProps
+        }
+        else
+            return new Properties()
+    }
+
+    static String getLabKeyServer(Project project)
+    {
+        return getTeamCityProperty(project, "labkey.server", "http://localhost")
+    }
+
+    static String getLabKeyContextPath(Project project)
+    {
+        return getTeamCityProperty(project, "labkey.contextpath", null)
+    }
+
+    static String getLabKeyServerPort(Project project)
+    {
+        return getTeamCityProperty(project, 'tomcat.port', null)
+    }
```

Comment:
Putting these static TeamCity property getters in the task class seems odd. It also causes some method name overlap; there is a static `getLabKeyServerPort` that gets the property but there is also the implicit non-static `getLabKeyServerPort` for the task input.
If `TeamCityExtension` isn't the right place for these, maybe they could live in a new `TeamCityUtils` (similar to `BuildUtils`).

Reply:
The placement of these methods doesn't seem odd to me, but good point about the name overlap. I don't remember the evolution of this change, but I am generally moving away from heavy usage of Extension classes where possible because we don't seem to be using them in the way they are intended in a lot of places. No real harm for static classes, though, so I've moved these methods back (for now, at least).